keps/0032-community-repository-management.md

```diff
@@ -0,0 +1,96 @@
+
+## Motivation
+
+Currently, every operator package in the community repository is provided from the `kudobuilder/operators` Git repository. This Git repository contains package definitions, documentation and sometimes tests for each operator. While it is convenient, to have this all in a single repository, this is challenging for larger or third-party operators:
+
+- Large operators usually provide own Docker images and/or additional dependencies. These have to be build, tested and deployed. The tools for this can't be provided by `kudobuilder/operators`.
+- Some operator tests cannot be covered by the test tooling provided in `kudobuilder/operators` making it necessary to host the tests in a separate Git repository.
+- Providing multiple versions of an operator is only possible by using separate folders. Teams developing operators might prefer different workflows (e.g. Git tags)
+
+These requirements make it necessary for operator packages to be developed in a separate Git repository. As a result, the operator packages in `kudobuilder/operators` are copies of specific versions of the respective upstream Git repository. This approach has challenges as well, because bugs discovered in an operator need to be resolved in the upstream Git repository, not in `kudobuilder/operators`. I.e., there needs to be metadata to link to the upstream Git repository. By having some form of metadata describing upstream operator package sources, the same metadata can be used to describe other properties of a package, e.g. the maturity level.
+
+### Goals
+
+- Provide a simple workflow for upstream operator packages to get added to the community repository
+- Remove the need to "host" copies of upstream operator packages in a Git repository
+- Provide a mechanism to define upstream operator package sources
+- Provide a mechanism to define upstream operator package maturity
+
+### Non-Goals
+
+- Change the practice of creating a PR against a Git repository to add operator packages to the community repository
+- Change operator package management as defined in [KEP-10](0010-package-manager.md) and [KEP-15](0015-repository-management.md)
+
+## Proposal
+
+Upstream operator developers still create PRs against a Git repository to add their operator packages to the community repository. This Git repository lists references to upstream operator packages instead of a full copy of an operator package. For example, consider an operator package developed at `github.com/example/example-operator` that has a tagged version `1.0.0` and the operator package in the `operator` folder. To add or update this operator package, the developers would create a PR referencing their upstream Git repository and the specific version, e.g. by adding a file `example-operator.yaml` like
+
+```yaml
+name: Example Operator
+versions:
+- version: "1.0.0"
+  git:
+    repository: github.com/example/example-operator.git
+    tag: "1.0.0"
+    dir: operator
+```
+
```

Comment:
Should we be responsible for compiling the operator as part of the KUDO community repo, or should we be referencing releases that are built and hosted on GitHub (e.g. like Krew does https://github.com/kubernetes-sigs/krew-index/blob/master/plugins/kudo.yaml)?

By referencing a release, we could eventually support various release objects (e.g. helm charts/Docker Images/OCI Artifact Specs/etc) rather than requiring the internal objects to build those objects.

Reply:
It depends on what "compiling" should cover. We should be responsible for creating tarballs from operator package folders because these tarballs are specific to repositories and operator developers shouldn't keep them as part of their Git repositories. Though we could also support these tarballs directly if developers want to do that.

Reply:
Agreed. Creating the tarballs ourself also makes sure that they're not changed afterwards

Reply:
I'm ok with having a model that creates the tarballs... but I also want to support referencing prebuilt tars.   Requiring us to build them assumes that the build process is exactly how we define it... and that the git repo has an instance of this...  I could imagine additional templates or build techniques for which the files to create the operator are an intermediary step (which an org may not want to check-in, outside of our requirement on them).   Supporting prebuilts is necessary IMO.   We should: 1) either make this a current non-goal (for future definition) or 2) define it.

Reply:
The use of "version" concerns me a bit... 

context:  The version of an operator is defined in the operator.yaml and is part of the tarball name.   The tarball version also defines this as app_version and op_version.   so... 

1. Is the "version" used here... app, or op?
2. Is this expected to be the app_V-op_v?
3. Is it checked / verified against the artifact?  does the PR fail if it doesn't match?

Additionally... can we make it easy for people... can we assume that the "tag" is the "version" if not specified?  It seems like we could assume that the version is the tag but allow tag to override that assumption.

Reply:
Each package that is to be added to the community repo needs to be checked for _validity_ by the CI. This would at least cover that it is recognized by KUDO as an operator package. More checks for conformance are of course possible as well. All of these checks will work with directories as well as tarballs. When referencing tarballs, these will be copied into the community repository. I'll clarify this in the KEP.

@kensipe Good point regarding the version. This is again a question on which metadata should in these reference files versus what metadata is provided by the actual operator package. A reference file could be as simple as
```
versions:
-   git:
    repository: github.com/example/example-operator.git
    tag: "1.0.0"
    dir: operator
```
because the other metadata (`name` as well as `version`) is redundant. But IMO there's value to have them duplicated here, as it helps developers to understand what is being referenced. These values will show up in CI logs and could help debug CI failures, e.g. when using a wrong Git tag.